src/SUnitRange.jl

```diff
@@ -30,3 +30,31 @@ show(io::IO, ::Type{SUnitRange}) = print(io, "SUnitRange")
 function show(io::IO, ::MIME"text/plain", ::SUnitRange{Start, L}) where {Start, L}
     print(io, "SUnitRange($Start,$(Start + L - 1))")
 end
+
+# For this type to be usable as `indices`, they need to support some more stuff
+Base.unsafe_length(r::SUnitRange) = length(r)
+@inline first(r::SUnitRange{Start}) where {Start} = Start # matches Base.UnitRange when L == 0...
+@inline endof(r::SUnitRange{Start, L}) where {Start, L} = L
+
+(==)(::SUnitRange{Start, L}, ::SUnitRange{Start, L}) where {Start, L} = true
+(==)(::SUnitRange{Start1, L1}, ::SUnitRange{Start2, L2}) where {Start1, Start2, L1, L2} = false
+
+(==)(::SUnitRange{1, L}, r::Base.OneTo) where {L} = L == r.stop
+(==)(::SUnitRange, ::Base.OneTo) = false
+(==)(r::Base.OneTo, ::SUnitRange{1, L}) where {L} = L == r.stop
+(==)(::Base.OneTo, ::SUnitRange) = false
+
+start(r::SUnitRange) = 1
+next(r::SUnitRange, i) = (r[i], i+1)
+done(r::SUnitRange{Start, L}, i) where {Start, L} = i > L
+
+@pure Base.UnitRange{Int}(::SUnitRange{Start, L}) where {Start, L} = Start : (Start + L - 1)
+@pure Base.Slice(::SUnitRange{Start, L}) where {Start, L} = Base.Slice(Start : (Start + L - 1)) # TODO not optimal
```

Comment:
This is not wrong but is perhaps pushing the limits. `Base.Slice` is a (deliberately) limited version of [IdentityRange](https://github.com/JuliaArrays/IdentityRanges.jl), and we only say that two arrays are equal if they have *both* the same values and the same indices. Since this is a constructor rather than a conversion, you might be able to get away with this, but I'd also add
```julia
convert(::Type{Base.Slice}, r::SUnitRange{1,L}) = Base.Slice(r)
convert(::Type{Base.Slice}, r::SUnitRange) = error("$r cannot be converted to Base.Slice, consider `Base.Slice(r)` instead")
```
(`convert` has an implicit promise of preserving at least approximate equality relationships, see the many discussions about whether it's OK to strip physical units e.g., `convert(Int, 3meter)`.)